File: 08-Seven-Segment-Search.js
const R = require('ramda')
const readline = require('readline')
const rl = readline.createInterface({ input: process.stdin })

const lines = []

rl.on("line", data => {
  const [signalPatterns, outputValues] = data.split(' | ').map(v=>v.split(' '))

  lines.push({signalPatterns, outputValues})
})

rl.on('close', () => {
  console.log('partOne', partOne(lines))
  console.log('partTwo', partTwo(lines))
})

function partOne(lines) {
  let sum = 0
  for (const {outputValues} of lines) {
    sum+=outputValues.filter(v => [2,3,4,7].includes(v.length)).length
  }

  return sum
}

const includes = (str, substr) => R.difference(substr, str).length === 0
const sortStr = str => str.split('').sort().join('')
const toSortedArrays = str => str.split('').sort()

function analyze(signalPatterns) {
  const patterns = signalPatterns.map(toSortedArrays)

  const one   = patterns.find(s => s.length === 2)
  const four  = patterns.find(s => s.length === 4)
  const seven = patterns.find(s => s.length === 3)
  const eight = patterns.find(s => s.length === 7)
  const three = patterns.find(s => s.length === 5 && includes(s, one))
  const six   = patterns.find(s => s.length === 6 && (!s.includes(one[0]) || !s.includes(one[1])))
  const nine  = patterns.find(s => s.length === 6 && includes(s, three))
  const zero  = patterns.find(s => s.length === 6 && s !== six && s !== nine)
  const five  = patterns.find(s => s.length === 5 && s !== three && includes(nine, s))
  const two   = patterns.find(s => s.length === 5 && s !== three && s !== five)

  const analysis = {
    zero, one, two, three, four, five, six, seven, eight, nine
  }

  return Object.values(analysis).reduce((dict, v,i) => {dict[v.join('')] = i; return dict}, {})
}


function partTwo(lines) {
  let  sum = 0
  for (const {signalPatterns, outputValues} of lines) {
    const dict = analyze(signalPatterns)
    const num = outputValues.map(s => dict[sortStr(s)])
    sum += Number.parseInt(num.join(''))
  }
  return sum
}

File: 2021/08-Seven-Segment-Search.hs
module SevenSegmentSearch where

import Data.List.Split (splitOn)
import Data.List (sort, elemIndex)
import Data.Maybe
import Utils (fromDigits)

type SignalPatterns = [String]
type OutputValues = [String]

parseInput :: [String] -> [(SignalPatterns, OutputValues)]
parseInput = map (tup . splitOn " | ")
  where
    tup [x, y] = (words x, words y)

find :: (a -> Bool) -> [a] -> a
find f = head . filter f

includes :: String -> String -> Bool
includes str = all (`elem` str)

elemIndex' :: [String] -> String -> Int
elemIndex' xs x = fromMaybe (-1) $ elemIndex x xs -- dit gaat stilletje heel erg mis...

-- | put the signal patterns in order
analyze :: SignalPatterns -> SignalPatterns
analyze signals = [zero, one, two, three, four, five, six, seven, eight, nine]
  where
    one   = find ((==2) . length) signals
    four  = find ((==4) . length) signals
    seven = find ((==3) . length) signals
    eight = find ((==7) . length) signals
    three = find (\s -> length s == 5 && includes s one) signals
    six   = find (\s -> length s == 6 && (notElem (head one) s  || notElem (last one) s))  signals
    nine  = find (\s -> length s == 6 && includes s three) signals
    zero  = find (\s -> length s == 6 && s /= six   && s /= nine) signals
    five  = find (\s -> length s == 5 && s /= three && includes nine s) signals
    two   = find (\s -> length s == 5 && s /= three && s /= five) signals

partOne :: [(SignalPatterns, OutputValues)] -> Int
partOne signals = sum $ map matches signals
  where matches (_,output) = length $ filter (\o -> length o `elem` [2,3,4,7]) output

partTwo :: [(SignalPatterns, OutputValues)] -> Int
partTwo signals = sum $ map (fromDigits . decode) signals
 where
   decode (patterns, output) = map (elemIndex' (dict patterns) . sort) output
   dict = map sort . analyze

main = do
  contents <- getContents
  let input = lines contents
      signals = parseInput input

  print $ partOne signals -- 532
  print $ partTwo signals -- 1011284
